### ISSUE TITLE
fix(lora): refuse adapters whose tensors do not map onto the base model instead of serving the base weights

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#1328]
## Summary

`mlxcel-server --adapter <dir>` (and `mlxcel generate --adapter`) logs "Fused LoRA adapters into N layers" where `N` counts `.lora_a` tensors in the adapter file, not pairs that were actually fused. A pair whose base weight is missing (adapter trained for a different architecture, a renamed projection, a typo in the path's `adapters.safetensors`) is skipped with a `warn!` and the server starts and serves the unmodified base model while claiming the adapter is loaded. A DoRA adapter is accepted as LoRA and its magnitude vectors are silently dropped, producing weights that match neither the base nor the fine-tune. This change makes adapter loading fail with a listing of every tensor that could not be applied, reports the true fused count, and refuses DoRA until its fusion is implemented.

## Current behavior

`src/lora/loader.rs:129-139`: tensors not ending in `.lora_a` / `.lora_b` are ignored with the comment `// Ignore other weights (like scales for DoRA)`.

`src/lora/loader.rs:150-168`:

```rust
let (Some(lora_a), Some(lora_b)) = (lora_a_opt, lora_b_opt) else {
    tracing::warn!("Incomplete LoRA pair for {}: missing lora_a or lora_b", base_name);
    continue;
};
let base_weight_name = find_base_weight_name(&base_name, base_weights)?;
let Some(base_weight) = base_weights.get(&base_weight_name) else {
    tracing::warn!("Base weight not found for LoRA layer {}: tried {}", base_name, base_weight_name);
    continue;
};
```

`find_base_weight_name` (`src/lora/loader.rs:353-373`) never fails: when no candidate exists it returns `"{name}.weight"`, which then hits the `warn!` above.

`src/lora/loader.rs:469-476` (`apply_lora_adapters`) and `src/lora/loader.rs:523-540` (`apply_stage_lora_adapter`) compute `modified_count` from adapter keys before fusion and log it as the number of fused layers.

`src/lora/config.rs:139-141`: `is_lora()` returns true for `FineTuneType::DoRA`, so `apply_lora_adapters` accepts a DoRA adapter and fuses only its `lora_a`/`lora_b`.

Shape mismatches are already hard errors (`src/lora/loader.rs:183-199`), and the GPT-2 transposed-layout guard is a hard error (`reject_conv1d_layout_fusion`). Only the missing-base-weight and unexpected-tensor cases are silent.

The server wires the flag through `src/server/model_worker.rs:252-254` (`backend.load_model_with_adapter`) into `src/loading/mod.rs:701` (`lora::apply_lora_adapters`); the pipeline-parallel path uses `apply_stage_lora_adapter` from `src/distributed/pipeline/stage_executor/llama.rs:60` with a per-stage `LayerFilter`.

## Expected behavior

Loading an adapter directory validates every tensor in `adapters.safetensors` (or `adapter_model.safetensors`) before any fusion:

1. Every tensor name must end in `.lora_a` or `.lora_b` (after the HuggingFace PEFT `.base_layer` normalization already handled by `find_base_weight_name`). Any other leaf (`.m`, `.scales`, `.weight`, ...) is an error naming the tensor.
2. Every `.lora_a` must have a matching `.lora_b` and vice versa; an incomplete pair is an error naming the layer.
3. Every pair's base weight must exist in the base `WeightMap` under one of the `find_base_weight_name` candidates; otherwise an error naming the layer and the candidates tried.
4. `lora_a` and `lora_b` must be rank-compatible in one of the two accepted orientations (already enforced by `compute_lora_delta`), and the delta must match the base weight shape (already enforced).
5. `fine_tune_type: "dora"` is rejected: `"DoRA adapters are not supported: <path> declares fine_tune_type dora; fuse it with its own tooling first"`. `"full"` stays rejected as today.

All violations are collected and reported in one error:

```
Adapter at <dir> contains tensors that cannot be applied to this model:
  model.layers.0.self_attn.q_proj.lora_a: no base weight (tried model.layers.0.self_attn.q_proj.weight, model.layers.0.self_attn.q_proj)
  model.layers.3.mlp.down_proj.m: not a LoRA tensor (expected .lora_a or .lora_b)
```

The "fused into N layers" log reports the number of pairs actually added into `base_weights`.

Pipeline stages: `apply_stage_lora_adapter` validates only the tensors inside its `LayerFilter` (tensors for other stages are never loaded). A stage whose range contains no adapter tensors is valid (an adapter may target a subset of layers); an in-range tensor that fails rules 1-4 is an error exactly as above.

## Implementation plan

1. `src/lora/loader.rs`: add

   ```rust
   /// Pair up adapter tensors and check each against the base weights.
   /// Returns the fusable pairs keyed by base weight name, or one error
   /// listing every offending tensor.
   fn validate_adapter_tensors(
       base_weights: &WeightMap,
       adapter_weights: &WeightMap,
   ) -> Result<Vec<(String /* base weight key */, UniquePtr<MlxArray> /* a */, UniquePtr<MlxArray> /* b */)>>
   ```

   implementing rules 1-3. Make `find_base_weight_name` return `Option<String>` (remove the "most likely candidate" fallback) and have the validator format the candidate list on `None`.

2. Rewrite the first half of `fuse_lora_weights_into` to call `validate_adapter_tensors` and iterate the returned pairs; keep `reject_conv1d_layout_fusion`, `compute_lora_delta` and the shape guard unchanged. Return `Ok(fused_count)` (`Result<usize>`) and have `fuse_lora_weights` forward it.

3. `apply_lora_adapters` and `apply_stage_lora_adapter`: drop the key-count `modified_count`; log the `usize` returned by fusion. In `apply_lora_adapters`, error if the returned count is `0` (an adapter with no applicable tensor). Do not add that zero check to the stage path.

4. `src/lora/config.rs`: add `pub fn is_fusable_lora(&self) -> bool { self.fine_tune_type == FineTuneType::LoRA }` and use it in both `apply_*` entry points with the DoRA-specific message from Expected behavior step 5. Keep `is_lora()` for any other caller (grep first; remove it if unused).

5. `src/distributed/pipeline/partial_loading_adapter.rs` (`load_stage_adapter_weights`): no change needed if it returns a `WeightMap` of in-range tensors; confirm non-LoRA leaves inside the range are retained so the validator can report them.

## Validation

(a) Unit tests, `src/lora/loader_tests.rs` (reuse `lora_pair(layer, in, rank, out)` at line 38):

- `adapter_pair_without_base_weight_is_an_error`: base has `layer.weight`; adapter is `lora_pair("other", 4, 2, 3)`; `fuse_lora_weights` returns `Err` whose message contains `other.lora_a` and `tried other.weight`.
- `adapter_with_unknown_leaf_is_an_error`: adapter `lora_pair("layer", ...)` plus `layer.m` ones `[3]`; `Err` mentions `layer.m` and `not a LoRA tensor`.
- `incomplete_pair_is_an_error`: only `layer.lora_a`; `Err` mentions `layer.lora_b`.
- `all_violations_are_reported_together`: two bad layers; both names appear in the single error string.
- `fused_count_reports_applied_pairs`: two valid pairs on two base weights; `fuse_lora_weights_into` returns `Ok(2)`.
- `dora_adapter_is_refused`: write a temp dir with `adapter_config.json` `{"fine_tune_type": "dora", "lora_parameters": {"rank": 8, "scale": 20.0}}` and an `adapters.safetensors` holding one valid pair; `apply_lora_adapters` returns `Err` containing `DoRA`.
- `stage_with_no_in_range_adapter_tensors_is_valid`: `apply_stage_lora_adapter` with a `LayerFilter` covering layers the adapter does not touch returns `Ok(())` and fuses `0`.
- Existing tests (`test_fuse_lora_weights_basic`, the Conv1D guards, the transpose-hint tests) keep passing.

(b) Real checkpoint: `mlx-community/Meta-Llama-3.1-8B-Instruct-4bit` (local `models/meta-llama-3.1-8b-instruct-4bit`) with a locally trained rank-8 LoRA adapter on 4 layers (any short fine-tune run; the adapter only needs a valid `adapter_config.json` with `fine_tune_type: lora`, `lora_parameters: {rank, scale}` and an `adapters.safetensors`).

```
./target/release/mlxcel generate -m models/meta-llama-3.1-8b-instruct-4bit --adapter <adapter-dir> -p "Hello" -n 32
./target/release/mlxcel-server -m models/meta-llama-3.1-8b-instruct-4bit --adapter <adapter-dir> --port 8080
```

Acceptance: the log line reads `Fused LoRA adapters into 16 layers` (4 layers x q/k/v/o, or whatever the adapter's `lora_layers` x target modules product is) and generation differs from the base model on the same prompt. Then copy the adapter and rename one tensor to `model.layers.0.self_attn.q_proj.lora_a.bogus`: both commands must exit non-zero with the tensor name in the error, and the server must not start. Then point `--adapter` at an adapter trained for `mlx-community/Qwen2.5-7B-Instruct-4bit` (local `models/qwen2.5-7b-instruct-4bit`) while loading the Llama checkpoint: startup fails listing the unmatched layers.

## Acceptance criteria

- [x] An adapter with any tensor that cannot be applied (unknown leaf, incomplete pair, missing base weight) fails to load with every offender listed, on both the single-process and pipeline-stage paths
- [x] The fused-layer count in the logs equals the number of pairs actually added
- [x] DoRA adapters are refused with a message naming DoRA
- [x] Valid adapters load exactly as before (same fused weights)
- [x] cargo test --profile test-fast --features metal,accelerate --lib lora / --lib distributed::pipeline / --lib loading:: pass (61 / 323 / 312). The full `--workspace` run was left to CI and to the orchestrator, not run locally.
- [x] cargo clippy --profile test-fast --lib --tests --features metal,accelerate -- -D warnings and cargo fmt --all -- --check pass. The `--workspace --all-targets` clippy variant was left to CI.

## Out of scope

- Implementing DoRA fusion (`W' = m * (W + scale * B A) / ||W + scale * B A||` per output row); a follow-up once a DoRA adapter checkpoint is available to validate against.
- Runtime (unfused) LoRA application through `mlxcel_core::layers::LoRAWeights`.

Implemented in PR #1576. Two deviations from the plan above, both deliberate: `find_base_weight_name` became private (its only remaining callers are inside `loader.rs`), and the unfused runtime staging path in `src/lora/runtime.rs` adopted the same validator, because it carried the identical warn-and-continue posture with a source comment assigning that work to this issue and it is the default channel for the b10621 `--lora` spellings.
